### DISCUSSION TITLE
User holding down key - handling

### DISCUSSION THREAD

**Discussion by John-Nagle** [rust-windowing/winit#2383] [Q&A]
For game purposes and for arrow keys, what's the proper Winit way to handle key repeats? Winit will generate a Pressed and Released event. Do I have to track which keys are down, and for how long, or is there something in Winit which will generate repeated events for held-down keys?

**Comment by maroider** [comment]
"True" `Pressed` and `Released` events always come in pairs (that's not *quite* true, but it's true enough). In the case of key repeat, you'll receive consecutive `Pressed` events for a given key without any `Released` event in-between. Thus, you should consider any `Pressed` events after the first to be repeat events. If you don't receive any repeat events, then could possibly be because the underlying API doesn't send any (as *may* be the case for `DeviceEvent`s).

**Comment by John-Nagle** [comment]
OK. I ended up keeping a HashSet<VirtualKeyCode> to track which keys are currently down. It's updated as key events come in. On each scene frame, I check that set to decide what movement is appropriate. It's a workable solution.

**Winit** knows to keep re-sending "Pressed", so it must have some internal info like that.  If **winit** has that info, I'd like to have access to it, rather than duplicating that work.

**Reply by maroider** [reply]
> If winit has that info, I'd like to have access to it, rather than duplicating that work.

I think we do on some platforms, but not all.
